File: algorithms/java/src/test/java/leetcode/LC341FlattenNestedListIterator.java
package leetcode;

import org.junit.jupiter.api.Test;


import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Stack;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/flatten-nested-list-iterator/
 * #medium
 */
public class LC341FlattenNestedListIterator {
    public class NestedInteger {
        private Object object;

        public NestedInteger(Object object) {
            this.object = object;
        }

        public boolean isInteger() {
            return object instanceof Integer;
        }

        public Integer getInteger() {
            return (Integer) object;
        }

        public List<NestedInteger> getList() {
            return (List<NestedInteger>) object;
        }

        public String toString() {
            if (object instanceof Integer) {
                return object.toString();
            } else {
                List<NestedInteger> list = (List<NestedInteger>) object;
                StringBuilder builder = new StringBuilder();
                builder.append("(");
                for (NestedInteger nested : list) {
                    builder.append(nested.toString());
                }
                builder.append(")");
                return builder.toString();
            }
        }
    }

    public class NestedIterator implements Iterator<Integer> {
        private class Item {
            public Object item;
            public int skip;

            public Item(Object item, int skip) {
                this.item = item;
                this.skip = skip;
            }

            public String toString() {
                return String.format("item: %s; skip: %d", item, skip);
            }
        }

        private Stack<Item> stack = new Stack<>();

        public NestedIterator(List<NestedInteger> nestedList) {
            stack.push(new Item(nestedList, 0));
        }

        private Object objectify(NestedInteger nested) {
            if (nested.isInteger()) {
                return nested.getInteger();
            } else {
                return nested.getList();
            }
        }

        private void find() {
            if (stack.isEmpty()) return;
            while (!stack.isEmpty() && !(stack.peek().item instanceof Integer)) {
                Item top = stack.peek();
                List<NestedInteger> list = (List<NestedInteger>) top.item;
                if (list.size() <= top.skip) {
                    stack.pop();
                    continue;
                }
                stack.push(new Item(objectify(list.get(top.skip)), 0));
                top.skip++;
            }
        }

        @Override
        public Integer next() {
            if (hasNext()) {
                return (Integer) stack.pop().item;
            } else {
                return null;
            }
        }

        @Override
        public boolean hasNext() {
            find();
            return !stack.isEmpty();
        }
    }

    @Test
    public void test_example_1() throws Exception {
        List<NestedInteger> list_1_1_a = new ArrayList<>();
        list_1_1_a.add(new NestedInteger(1));
        list_1_1_a.add(new NestedInteger(1));
        List<NestedInteger> list_1_1_b = new ArrayList<>();
        list_1_1_b.add(new NestedInteger(1));
        list_1_1_b.add(new NestedInteger(1));
        List<NestedInteger> list = new ArrayList<>();
        list.add(new NestedInteger(list_1_1_a));
        list.add(new NestedInteger(2));
        list.add(new NestedInteger(list_1_1_b));
        int[] expected = {1, 1, 2, 1, 1};
        NestedIterator nested = new NestedIterator(list);
        for (int value : expected) {
            assertEquals(true, nested.hasNext());
            assertEquals(value, nested.next().intValue());
        }
        assertEquals(false, nested.hasNext());
        assertEquals(null, nested.next());
    }

    @Test
    public void test_example_2() throws Exception {
        List<NestedInteger> list = new ArrayList<>();
        list.add(new NestedInteger(1));
        List<NestedInteger> list_4 = new ArrayList<>();
        list_4.add(new NestedInteger(4));
        List<NestedInteger> list_6 = new ArrayList<>();
        list_6.add(new NestedInteger(6));
        list_4.add(new NestedInteger(list_6));
        list.add(new NestedInteger(list_4));
        int[] expected = {1, 4, 6};
        NestedIterator nested = new NestedIterator(list);
        for (int value : expected) {
            assertEquals(true, nested.hasNext());
            assertEquals(value, nested.next().intValue());
        }
        assertEquals(false, nested.hasNext());
        assertEquals(null, nested.next());
    }
}

File: algorithms/code/leetcode/lc341_flatten_nested_list_iterator/lc341_flatten_nested_list_iterator.rs
// https://leetcode.com/problems/flatten-nested-list-iterator/
// #medium

#[derive(Debug, PartialEq, Eq)]
pub enum NestedInteger {
    Int(i32),
    List(Vec<NestedInteger>),
}

pub struct NestedIterator {
    flattened: Vec<i32>,
    index: usize,
}

#[allow(clippy::should_implement_trait)]
impl NestedIterator {
    pub fn new(nested_list: Vec<NestedInteger>) -> Self {
        let mut flattened = Vec::new();
        Self::flatten(&nested_list, &mut flattened);
        Self {
            flattened,
            index: 0,
        }
    }

    fn flatten(nested_list: &[NestedInteger], flattened: &mut Vec<i32>) {
        for nested_integer in nested_list {
            match nested_integer {
                NestedInteger::Int(value) => flattened.push(*value),
                NestedInteger::List(list) => Self::flatten(list, flattened),
            }
        }
    }

    pub fn next(&mut self) -> i32 {
        let value = self.flattened[self.index];
        self.index += 1;
        value
    }

    pub fn has_next(&self) -> bool {
        self.index < self.flattened.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn generic(list: Vec<NestedInteger>, expected: Vec<i32>) {
        let mut nested = NestedIterator::new(list);
        for value in expected {
            assert!(nested.has_next());
            assert_eq!(value, nested.next());
        }
        assert!(!nested.has_next());
    }

    #[test]
    fn test_example_1() {
        let mut list11a = Vec::new();
        list11a.push(NestedInteger::Int(1));
        list11a.push(NestedInteger::Int(1));
        let mut list11b = Vec::new();
        list11b.push(NestedInteger::Int(1));
        list11b.push(NestedInteger::Int(1));
        let mut list = Vec::new();
        list.push(NestedInteger::List(list11a));
        list.push(NestedInteger::Int(2));
        list.push(NestedInteger::List(list11b));
        let expected = vec![1, 1, 2, 1, 1];
        generic(list, expected);
    }

    #[test]
    fn test_example_2() {
        let mut list = Vec::new();
        list.push(NestedInteger::Int(1));
        let mut list4 = Vec::new();
        list4.push(NestedInteger::Int(4));
        let mut list6 = Vec::new();
        list6.push(NestedInteger::Int(6));
        list4.push(NestedInteger::List(list6));
        list.push(NestedInteger::List(list4));
        let expected = vec![1, 4, 6];
        generic(list, expected);
    }
}
